### PR TITLE
Refactor/40 implement architecture by refactoring code into classes

### LINKED ISSUE #40: Implement architecture by refactoring code into classes
As a software developer, I want to have a modular architecture, so that I can implement future features more easily.

**Description:** already existent code should be refactored into classes by implementing the presented architecture

**Dependent Issues:** #37 

# Acceptance Criteria
- [x] all existent code is part of a class
- [x] every module has a corresponding class
- [x] the pipeline, as is it is at the beginning of working on this issue, works as intended
- [x] common requirements regarding modularity and maintenance are fulfilled

### PR DESCRIPTION
**Pull request by FR-SON**
This PR changes the basic structure of the project by introducing a more modular architecture featuring an orchestrator, that manages separate modules, each of which is responsible for one dedicated step in the pipeline.
Modules can now easily be added or replaced by following theses steps
1. Create a class derived from the `Module` class from `module.py`
2. import the class in `orchestrator.py`
3. Make the required changes to the dictionary `modules` in the `ExtractionConfiguration` class (adding, removing)
4. Make the same changes to the list in `initialize_modules()` in the `orchestrator` class
5. Keep in mind, that modules can depend on each other and choose the order accordingly

Currently there are two modules for extracting timestamps. Pick whichever you prefer and follow the steps above to use the chose module.

This closes #40.

### CODE REVIEW

File: requirements.txt [hunk 1]
```diff
@@ -2,3 +2,4 @@ openai==1.5.0
 pm4py
 django
 pre-commit
+pandas
```
Comment (nils-schmitt) [inline requirements.txt:5]:
```suggestion

```

File: tracex/extraction/logic/__init__.py [hunk 2]
```diff
@@ -0,0 +1 @@
+"""Contains all the logic for the extraction process."""
```
Comment (nils-schmitt) [inline tracex/extraction/logic/__init__.py:1]:
```suggestion

```

File: tracex/extraction/logic/modules/__init__.py [hunk 3]
```diff
@@ -0,0 +1 @@
+"""All modules used for the extraction, derived from the Module class."""
```
Comment (nils-schmitt) [inline tracex/extraction/logic/modules/__init__.py:1]:
```suggestion

```

File: requirements.txt (hunk 1 above)
Comment (nils-schmitt) [reply requirements.txt:5]: the pandas import is included in the pm4py import

File: tracex/extraction/prototype/pipeline_without_user_io.py
Comment (nils-schmitt) [inline tracex/extraction/prototype/pipeline_without_user_io.py:1]: This file should be deleted

**Review by nils-schmitt — CHANGES_REQUESTED** [review]

**Comment by tkv29** [comment]
As disscued I have 2 suggestions:
1. Currently the orchestrator is initialized before the necessary config is created, therefore the config/instance variables of the orchestrator are none. Suggestion would be to initialize the orchestrator when a conifg is available and not before the config is created, so we only have an instance when needed.
2. Currently the orchestrator is directly accessing the results of the modules. I would suggest to implement a request return pattern, so the module will repsonse to the orchestrator with the results instead.

File: tracex/extraction/prototype/pipeline_without_user_io.py
Comment (FR-SON) [reply tracex/extraction/prototype/pipeline_without_user_io.py:1]: Done

**Comment by FR-SON** [comment]
> As disscued I have 2 suggestions:
> 
>     1. Currently the orchestrator is initialized before the necessary config is created, therefore the config/instance variables of the orchestrator are none. Suggestion would be to initialize the orchestrator when a conifg is available and not before the config is created, so we only have an instance when needed.
> 
>     2. Currently the orchestrator is directly accessing the results of the modules. I would suggest to implement a request return pattern, so the module will repsonse to the orchestrator with the results instead.

The second suggestion is applied. The modules now return their results instead of saving them in instance variables. 
The first suggestion, though reasonable, can hopefully be postponed, as it requires considerable effort. I suggest to merge the discussed branch and work on the life-cycle management of the orchestrator on a separate branch so everyone can use the updated architecture asap. Thoughts on that?

**Review by nils-schmitt — APPROVED** [review]
Looks great, nice job!

Regarding the previous comment from Frederic, i agree on updating the architecture first and then manage the orchestrator seperately.

**Review by tkv29 — APPROVED** [review]
lgtm, as mentioned above the changes will be delievered separately

**Pull request merged**